### PR TITLE
Add NonStickyEventExecutorGroup

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

We offer DefaultEventExecutorGroup as an EventExecutorGroup which return OrderedEventExecutor and so provide strict ordering of event execution. One limitations of this implementation is that each contained DefaultEventExecutor will always be tied to a single thread, which can lead to a very unbalanced execution as one thread may be super busy while others are idling.

Modifications:

- Add NonStickyEventExecutorGroup which can be used to wrap another EventExecutorGroup (like UnorderedThreadPoolEventExecutor) and expose ordering while not be sticky with the thread that is used for a given EventExecutor. This basically means that Threads may change between execution of tasks for an EventExecutor but ordering is still guaranteed.

Result:

Better utalization of threads in some use-cases.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler @trustin WDYT ?

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+            super(executor);
+            this.executor = executor;
+        }
+
+        @Override
+        public void run() {
+            if (!state.compareAndSet(SUBMITTED, RUNNING)) {
+                return;
+            }
+            if (tasks.isEmpty()) {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: just let `tasks.poll()` take care of this?

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+                return;
+            }
+            for (;;) {
+                Runnable task = tasks.poll();
+                if (task == null) {
+                    break;
+                }
+                task.run();
+            }
+            state.set(NONE);
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: put this in a `finally` block?

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+        public void execute(Runnable command) {
+            if (!tasks.offer(command)) {
+                throw new RejectedExecutionException();
+            }
+            if (state.compareAndSet(NONE, SUBMITTED)) {
+                // Actually it could happen that the runnable was picked up in between but we not care to much and just
+                // execute ourself. At worst this will be a NOOP when run() is called.
+                try {
+                    executor.execute(this);
+                } catch (RuntimeException e) {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: `RuntimeException` -> `Throwable` if we always want to remove the task

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+        @Override
+        public void run() {
+            if (!state.compareAndSet(SUBMITTED, RUNNING)) {
+                return;
+            }
+            if (tasks.isEmpty()) {
+                return;
+            }
+            for (;;) {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: couldn't a single thread still get hot if tasks are continuously executed? should we have some "max tasks per run" threshold where we re-execute this `Runnable`?

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    }
+
+    @Override
+    public EventExecutor next() {
+        return new NonStickyOrderedEventExecutor(group.next());
+    }
+
+    @Override
+    public Iterator<EventExecutor> iterator() {
+        final Iterator<EventExecutor> executors = group.iterator();
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: nit: `executors` -> `itr` (or something indicating this is an `iterator`)

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+import java.util.Queue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * {@link EventExecutorGroup} which wraps another {@link EventExecutorGroup} which contains <strong>NO</strong>
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]:
```
{@link EventExecutorGroup} which will preserve {@link Runnable} execution order but makes no guarantees about what {@link EventExecutor} (and therefore {@link Thread}) will be used to execute the {@link Runnable}s.

<p>The {@link EventExecutorGroup#next()} for the wrapped {@link EventExecutorGroup} must <strong>NOT</strong> return executors of type {@link OrderedEventExecutor}.
```

File: common/src/main/java/io/netty/util/concurrent/UnorderedThreadPoolEventExecutor.java [hunk 7]
```diff
@@ -37,7 +37,7 @@
  *
  * <strong>Because it provides no ordering care should be taken when using it!</strong>
  */
-public final class UnorderedThreadPoolEventExecutor extends ScheduledThreadPoolExecutor implements EventExecutor {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/UnorderedThreadPoolEventExecutor.java]: is this change necessary?

File: common/src/test/java/io/netty/util/concurrent/NonStickyEventExecutorGroupTest.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    private static void execute(EventExecutorGroup group, CountDownLatch startLatch) throws Throwable {
+        EventExecutor executor = group.next();
+        Assert.assertTrue(executor instanceof OrderedEventExecutor);
+        final AtomicReference<Throwable> cause = new AtomicReference<Throwable>();
+        final AtomicInteger last = new AtomicInteger();
+        int tasks = 10000;
+        final CountDownLatch latch = new CountDownLatch(tasks);
+        startLatch.await();
+
+        for (int i = 1 ; i < tasks + 1; i++) {
```
Comment (Scottmitch) [inline common/src/test/java/io/netty/util/concurrent/NonStickyEventExecutorGroupTest.java]: `i < tasks + 1` -> `i <= tasks`

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java (hunk 4 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: good idea..

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java (hunk 2 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: yep

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java (hunk 1 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: yep

File: common/src/main/java/io/netty/util/concurrent/UnorderedThreadPoolEventExecutor.java (hunk 7 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/UnorderedThreadPoolEventExecutor.java]: nope... let me revert

File: common/src/test/java/io/netty/util/concurrent/NonStickyEventExecutorGroupTest.java (hunk 8 above)
Comment (normanmaurer) [inline common/src/test/java/io/netty/util/concurrent/NonStickyEventExecutorGroupTest.java]: yes good idea.

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+            this.executor = executor;
+        }
+
+        @Override
+        public void run() {
+            if (!state.compareAndSet(SUBMITTED, RUNNING)) {
+                return;
+            }
+            try {
+                for (;;) {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: @Scottmitch I need to think how to best limit the for loop. Not sure yet atm

Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: let me know if I can help ... are you thinking we may need a solution more complicated than the following?

```java
try {
  int i = 0;
  for (; i < maxRunCount; ++i) {
    ..
  }
} finally {
  state.set(NONE);
  if (i == maxRunCount) {
    executor.execute(this); // This should share code with the Execute method below and cover the same exception cases, etc...
  }
}
```

**Comment by normanmaurer** [comment]
@Scottmitch PTAL

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+                            break;
+                        }
+                        safeExecute(task);
+                    }
+                } finally {
+                    if (i == maxTaskExecutePerRun) {
+                        try {
+                            state.set(SUBMITTED);
+                            executor.execute(this);
+                            return; // done
```
Comment (netkins) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: BLOCKER Remove this return statement from this finally block. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1143)

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+                            // Reset the state back to running as we will keep on executing tasks.
+                            state.set(RUNNING);
+                            tasks.remove(this);
+                            // if an error happened we should just ignore it and let the loop run again as there is not
+                            // much else we can do. Most likely this was triggered by a full task queue. In this case
+                            // we just will run more tasks and try again later.
+                        }
+                    } else {
+                        state.set(NONE);
+                        return; // done
```
Comment (netkins) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: BLOCKER Remove this return statement from this finally block. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1143)

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java (hunk 10 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: its fine.

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java (hunk 11 above)
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: its fine.

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+                } finally {
+                    if (i == maxTaskExecutePerRun) {
+                        try {
+                            state.set(SUBMITTED);
+                            executor.execute(this);
+                            return; // done
+                        } catch (Throwable ignore) {
+                            // Reset the state back to running as we will keep on executing tasks.
+                            state.set(RUNNING);
+                            tasks.remove(this);
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: is `this` ever expected to be in `tasks`?

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: What you mean ?

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+            this.executor = executor;
+            this.maxTaskExecutePerRun = maxTaskExecutePerRun;
+        }
+
+        @Override
+        public void run() {
+            if (!state.compareAndSet(SUBMITTED, RUNNING)) {
+                return;
+            }
+            for (;;) {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: the case where the loop provides value seems to be exceptional ... should we extract the contents of this method which runs tasks into another method and rely on recursion for this case? That way for the "typical" use case we don't have an extra loop statement.

```java
        @Override
        public void run() {
            if (!state.compareAndSet(SUBMITTED, RUNNING)) {
                return;
            }
            doRun();
        }

        private void doRun() {
            int i = 0;
            try {
                for (; i < maxTaskExecutePerRun; i++) {
                    Runnable task = tasks.poll();
                    if (task == null) {
                        break;
                    }
                    safeExecute(task);
                }
            } finally {
                if (i == maxTaskExecutePerRun) {
                    try {
                        state.set(SUBMITTED);
                        executor.execute(this);
                    } catch (Throwable ignore) {
                        // Reset the state back to running as we will keep on executing tasks.
                        state.set(RUNNING);
                        doRun();
                        // if an error happened we should just ignore it and let the loop run again as there is not
                        // much else we can do. Most likely this was triggered by a full task queue. In this case
                        // we just will run more tasks and try again later.
                    }
                } else {
                    state.set(NONE);
                }
            }
        }
```

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: @Scottmitch I really doubt it matters and I kind of like the current version more.

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java (hunk 12 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: why do we need this line ... I don't think we ever do `tasks.add(this)` (it would be strange if we did).

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: @Scottmitch ah now I understand... yes you are right we should remove it. Good catch!

**Comment by normanmaurer** [comment]
@Scottmitch @trustin ok to merge ?

File: common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+                        }
+                    } else {
+                        state.set(NONE);
+                        return; // done
+                    }
+                }
+            }
+        }
+
+        private static void safeExecute(Runnable task) {
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/NonStickyEventExecutorGroup.java]: i have seen this a few times before ... should we put this in some internal utility?

**Comment by Scottmitch** [comment]
1 followup comment and lgtm

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (fc14ca3)
